internal/metrics/metrics.go

```diff
@@ -0,0 +1,75 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	CertRestart 		string = "cert_restart"
+	RollingRestart 		string = "rolling_restart"
+	ScheduledRestart 	string = "scheduled_restart"
+	ManagedState		string = "managed"
+	UnmanagedState		string = "unmanaged"
+)
+
+var (
+	restartCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "eo_elasticsearch_cr_restart_total",
+			Help: "Total number of times the nodes restarted due to Cert Restart or Rolling Restart or Scheduled Restart.",
+		}, []string{"reason"})
+
+	esClusterManagementState = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "eo_elasticsearch_cr_cluster_management_state",
+			Help: "Number of Elasticsearch cluster that are in Managed state or Unmanaged state.",
+		}, []string{"state"})
+
+	metricList = []prometheus.Collector{
+		restartCounter,
+		esClusterManagementState,
+	}
+)
+
+//This function registers the custom metrics to the kubernetes controller-runtime default metrics.
+func RegisterCustomMetrics() {
+	for _, metric := range metricList {
+		metrics.Registry.MustRegister(metric)
+	}
+}
+
+//Increment the metric value by "1" when the node restarts due to cert.
+func IncrementRestartCounterCert() {
+	restartCounter.With(prometheus.Labels{
+		"reason": CertRestart,
+	}).Inc()
+}
+
+//Increment the metric value by "1" when the node restarts due to rolling.
+func IncrementRestartCounterRolling() {
+	restartCounter.With(prometheus.Labels{
+		"reason": RollingRestart,
+	}).Inc()
+}
+
+//Increment the metric value by "1" when the node is scheduled for cert restart or rolling restart.
+func IncrementRestartCounterScheduled() {
+	restartCounter.With(prometheus.Labels{
+		"reason": ScheduledRestart,
+	}).Inc()
+}
+
+//Sets the metric value to "n" when the ES Cluster Management State is Managed.
+func SetEsClusterManagementStateManaged(number int) {
+	esClusterManagementState.With(prometheus.Labels{
+		"state": ManagedState,
+	}).Set(float64(number))
+}
+
+//Sets the metric value to "n" when the ES Cluster Management State is Unmanaged.
+func SetEsClusterManagementStateUnmanaged(number int) {
+	esClusterManagementState.With(prometheus.Labels{
+		"state": UnmanagedState,
+	}).Set(float64(number))
+}
```

Comment:
Reflecting a little bit on the [prometheus/client-go gauge interface](https://github.com/prometheus/client_golang/blob/master/prometheus/gauge.go#L32), it looks to me that using `Set` is the wrong approach. We want do use `Add` for adding/removing clusters from each metric, i.e.:
- Cluster Managed -> Unmanaged: `Managed.Add(-1)` and `Unmanaged.Add(1)`
- Cluster Unmanager -> Managed: `Managed.Add(1)` and `Unmanaged.Add(-1)`

WDYT? (cc @blockloop)

Reply:
Indeed you are correct, `Set` will assign a constant value whatever we pass as param. Rather what we want here is to increment and decrement values whenever the cluster changes it's management set. `Add` helps us to achieve this. But the values might go negative as well. For example, initially both will be 0, when  we create a CR with `managed` state then `manage` will become 1 whereas `unmanaged` will become -1